### PR TITLE
[new release] grenier (0.12)

### PR DESCRIPTION
**Pull request by let-def**
A collection of various algorithms in OCaml

- Project page: <a href="https://github.com/let-def/grenier">https://github.com/let-def/grenier</a>
- Documentation: <a href="https://let-def.github.io/grenier/doc">https://let-def.github.io/grenier/doc</a>

##### CHANGES:

Balmap: alternative to Stdlib Maps and Sets based on baltree.
Dbseq: fast sequence datastructure for DeBruijn-indexed environments.
State elimination: convert e-NFA to regular expressions.

Fixed many bugs in and increased expressiveness of Valmari implementation.

### CODE REVIEW

**Comment by let-def** [comment]
Ok, some floating point heavy code is failing on x86-32 and arm64... I can look at that but I don't have access to such environments right now.

File: packages/grenier/grenier.0.12/opam [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+doc: "https://let-def.github.io/grenier/doc"
+build: [
+  ["dune" "subst"] {pinned}
+  ["dune" "build" "-p" name "-j" jobs]
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
+]
+depends: [
+  "ocaml" {>= "4.04"}
+  "dune" {>= "1.2.0"}
+]
```
Comment (mseri) [inline packages/grenier/grenier.0.12/opam:17]:
```suggestion
  "stdlib-shims"
]
```
You need to add stdlib-shims to the dependencies I think
```
#=== ERROR while compiling grenier.0.12 =======================================#
# context              2.0.8 | linux/x86_64 | ocaml-base-compiler.4.12.0 | pinned(https://github.com/let-def/grenier/releases/download/v0.12/grenier-v0.12.tbz)
# path                 ~/.opam/4.12/.opam-switch/build/grenier.0.12
# command              ~/.opam/4.12/bin/dune build -p grenier -j 47
# exit-code            1
# env-file             ~/.opam/log/grenier-16-a8fa88.env
# output-file          ~/.opam/log/grenier-16-a8fa88.out
### output ###
# File "balmap/dune", line 4, characters 28-40:
# 4 |  (libraries grenier.baltree stdlib-shims)
#                                 ^^^^^^^^^^^^
# Error: Library "stdlib-shims" not found.
# Hint: try:
#   dune external-lib-deps --missing -p grenier -j 47 @install
# File "strong/dune", line 6, characters 12-24:
# 6 |  (libraries stdlib-shims)
#                 ^^^^^^^^^^^^
# Error: Library "stdlib-shims" not found.
# Hint: try:
#   dune external-lib-deps --missing -p grenier -j 47 @install
```

**Comment by mseri** [comment]
The ocaml 4.13 failure is a dune failure: 
```
#=== ERROR while compiling dune.2.8.4 =========================================#
# context              2.0.8 | linux/x86_64 | ocaml-variants.4.13.0+trunk | file:///src
# path                 ~/.opam/4.13/.opam-switch/build/dune.2.8.4
# command              ~/.opam/4.13/bin/ocaml bootstrap.ml -j 47
# exit-code            2
# env-file             ~/.opam/log/dune-16-dc6e87.env
# output-file          ~/.opam/log/dune-16-dc6e87.out
### output ###
# ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot unix.cma boot/libs.ml boot/duneboot.ml
# ./.duneboot.exe -j 47
# cd _boot && /home/opam/.opam/4.13/bin/ocamlopt.opt -c -g -no-alias-deps -w -49 -I +threads dune_rules__cram_exec.ml
# File "src/dune_rules/cram_exec.ml", line 293, characters 50-61:
# Error: This expression has type sh_script
#        The field output_file does not belong to type sh_script
# 
```

File: packages/grenier/grenier.0.12/opam (hunk 1 above)
Comment (mseri) [reply packages/grenier/grenier.0.12/opam:18]: Can you upstream this change?

File: packages/grenier/grenier.0.12/opam [hunk 2]
```diff
@@ -7,7 +7,7 @@ license: "ISC"
 dev-repo: "git+https://github.com/let-def/grenier.git"
 doc: "https://let-def.github.io/grenier/doc"
 build: [
-  ["dune" "subst"] {pinned}
+  ["dune" "subst"] {dev}
```
Comment (mseri) [inline packages/grenier/grenier.0.12/opam:10]: Well spotted. @kit-ty-kate do you know if we can extend the lint action to check for this?

Comment (kit-ty-kate) [reply packages/grenier/grenier.0.12/opam:10]: the issue with adding a lint check for this is that half the packages that are autogenerated and will have this issue until they upgrade to dune >= 2.7. It's good to do for some packages but not for all (also given that most packages inside the opam-repository right now uses `{pinned}`)

File: packages/grenier/grenier.0.12/opam [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+doc: "https://let-def.github.io/grenier/doc"
+build: [
+  ["dune" "subst"] {dev}
+  ["dune" "build" "-p" name "-j" jobs]
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
+]
+depends: [
+  "ocaml" {>= "4.04"}
+  "dune" {>= "1.2.0"}
+  "stdlib-shims"
```
Comment (kit-ty-kate) [inline packages/grenier/grenier.0.12/opam:17]: Actually it seems that this library is not needed as the package already requires OCaml >= 4.08:
```
#=== ERROR while compiling grenier.0.12 =======================================#
# context              2.0.8 | linux/x86_64 | ocaml-base-compiler.4.06.1 | pinned(https://github.com/let-def/grenier/releases/download/v0.12/grenier-v0.12.tbz)
# path                 ~/.opam/4.06/.opam-switch/build/grenier.0.12
# command              ~/.opam/4.06/bin/dune build -p grenier -j 47
# exit-code            1
# env-file             ~/.opam/log/grenier-16-2fbc47.env
# output-file          ~/.opam/log/grenier-16-2fbc47.out
### output ###
#       ocamlc balmap/.balmap.objs/byte/balmap__Map.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I balmap/.balmap.objs/byte -I /home/opam/.opam/4.06/lib/stdlib-shims -I baltree/.grenier_baltree.objs/byte -no-alias-deps -open Balmap -o balmap/.balmap.objs/byte/balmap__Map.cmo -c -impl balmap/map.ml)
# File "balmap/map.ml", line 140, characters 12-23:
# Error: Unbound module Int
```
@let-def could you re-release by either increasing the minimal requirement for OCaml to 4.08 or changing the code that use the `Int` module? Also could you at the same time return the `{pinned} --> {dev}` fix?

Comment (let-def) [reply packages/grenier/grenier.0.12/opam]: Thanks @mseri and @kit-ty-kate. I have updated the constraints, only the floating point bugs remain.

**Comment by kit-ty-kate** [comment]
> Ok, some floating point heavy code is failing on x86-32 and arm64... I can look at that but I don't have access to such environments right now.

you can have access to an x86_32 environment using this dockerfile:
```
FROM ocaml/opam:debian-10-ocaml-4.12@sha256:4d6ccf85afe0755d6bf41a96cee16db0cfc5470a1f0249ab71591acff16f8bba
SHELL [ "/usr/bin/linux32", "/bin/sh", "-c" ]
```

**Comment by let-def** [comment]
Ok so, on 32 bits, the bug triggers only in native mode. In bytecode the test succeeds. This look a lot like a misbehaving optimisation, I am not sure I want to spend much time on that. Maybe I can restrict this library to x86-64 and investigate later why floating point code break some times ?

**Comment by kit-ty-kate** [comment]
Thanks!

**Pull request merged**